Mathlib/Probability/Entropy.lean

```diff
@@ -0,0 +1,304 @@
+/-
+Copyright (c) 2024 Rémy Degenne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rémy Degenne, Sébastien Gouëzel, Heather Macbeth
+-/
+import Mathlib.MeasureTheory.Constructions.Prod.Basic
+import Mathlib.Analysis.SpecialFunctions.Log.NegMulLog
+
+/-!
+# Entropy of a measure
+
+For a probability measure, the entropy is the quantity `-∑' s, (μ {s}).toReal * log (μ {s}).toReal`.
+
+## Main definitions
+
+* `measureEntropy`: entropy of a measure `μ`, denoted by `Hm[μ]`.
+
+## Notations
+
+* `Hm[μ] = measureEntropy μ`.
+
+-/
+
+open MeasureTheory Real BigOperators NNReal ENNReal
+
+namespace ProbabilityTheory
+
+variable {S T : Type*} {mS : MeasurableSpace S} {mT : MeasurableSpace T}
+  {μ : Measure S} {ν : Measure T}
+
+/-- Entropy of a measure. For a probability measure, this is the quantity
+`- ∑' s, (μ {s}).toReal * log (μ {s}).toReal`.
+Notation: `Hm[μ]`.
+
+We extend the definition to finite measures by normalizing the measure by `(μ Set.univ)⁻¹`.
+If `μ` is a probability measure, a call to `simp` will simplify `(μ Set.univ)⁻¹ • μ` to `μ`. -/
+noncomputable
+def measureEntropy (μ : Measure S := by volume_tac) : ℝ :=
+  ∑' s, negMulLog (((μ Set.univ)⁻¹ • μ) {s}).toReal
```

Comment:
Thanks for the suggestion. I am not really sure that we should normalize the measure anymore (partly because I did not record thoroughly enough the reason for normalizing in the first place) and I need to investigate that. But if normalizing is indeed a good idea, I'll look into using `cond`.

Reply:
This PR only introduces entropies of finite measures, right? Because even for probabilities, when they are continuous-valued you can have negative entropy (a.k.a. differential entropy).